HDDS-10775. Support bucket ownership verification

hevinhsu opened the pull request:
## What changes were proposed in this pull request?
This PR implements the feature described in [AWS S3 documentation on bucket owner condition](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-owner-condition.html#bucket-owner-condition-restrictions-limitations).
Since Ozone does not have an AWS Account ID (12-digit numeric value), the bucket owner's name is used for ownership validation instead.
The implementation covers the following operations:
* GetBucketAcl
* ListMultipartUpload
* ListObject 
* PutAcl
* HeadBucket
* DeleteBucket
* MultiDelete(Multi Object Delete)
* CreateKey
* PutObjectTagging
* CreateMultipartKey
* CreateMultipartByCopy
* CopyObject
* CreateDirectory
* GetKey
* GetObjectTagging
* ListPart
* HeadKey
* DeleteKey
* DeleteObjectTagging
* AbortMultipartUpload
* InitMultipartUplaod
* CompleteMultipartUpload
 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10775

## How was this patch tested?
This patch was tested using unit tests and integration tests.
https://github.com/hevinhsu/ozone/actions/runs/15431261432

hevinhsu commented on hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/awssdk/v2/AbstractS3BucketVerificationConditionUsingS3SDKV2Tests.java:
```diff
@@ -0,0 +1,618 @@
+ * - https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-owner-condition.html
+ */
+@TestMethodOrder(MethodOrderer.MethodName.class)
+public abstract class AbstractS3BucketVerificationConditionUsingS3SDKV2Tests extends OzoneTestBase {
```
That's a good idea. I'll move the test cases to AbstractS3SDKV2Tests.java.

hevinhsu commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketOwnerCondition.java:
```diff
@@ -0,0 +1,89 @@
+
+    // expectedSourceOwner is null, means the client does not want to check source owner
+    if (expectedSourceOwner != null && !sourceOwner.equals(expectedSourceOwner)) {
+      throw new OMException(ERROR_MESSAGE, OMException.ResultCodes.PERMISSION_DENIED);
```
According to the [AWS spec](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-owner-condition.html#bucket-owner-condition-when-to-use), when the verification fails, the expected response is ```Access Denied```.
I initially reused ```S3ErrorTable.ACCESS_DENIED``` mainly to leverage the [isAccessDenied](https://github.com/apache/ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/EndpointBase.java#L531-L535).

If needed, I'm open to adding a new error entry to ```S3ErrorTable``` for this specific case and updating the ```isAccessDenied``` method accordingly.

ivandika3 replied:
I think it will be better to create a new error entry to `S3ErrorTable` with the specific owner mismatch exception message (i.e. `BucketOwnerCondition.ERROR_MESSAGE`) and throw `OS3Exception` instead. While I understand that `OMException` will be caught in the outer scope and converted to `OS3Exception`, it seems odd to throw `OMException` when OM is not involved.

hevinhsu commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketOwnerCondition.java:
```diff
@@ -0,0 +1,89 @@
+   * @throws OMException if the expected bucket owner does not match the actual
+   *                     bucket owner
+   */
+  public static void verify(HttpHeaders headers, String bucketOwner) throws OMException {
```
Thanks for pointing this out. I initially thought owner would be a non-null value. I'll skip the verification when ```bucketOwner``` is null.

hevinhsu commented on hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/awssdk/v2/AbstractS3SDKV2Tests.java:
```diff
@@ -621,4 +648,512 @@ private void completeMultipartUpload(String bucketName, String key, String uploa
+    private static final String BUCKET_VERIFICATION_TEST_KEY = "test-key";
+
+    @BeforeAll
+    public void setup() throws Exception {
```
Thanks for providing the detailed test steps. I'll add both of these scenarios to the integration tests.

hevinhsu replied:
Hi @ivandika3 
I tested the dangling bucket scenario. After deleting the source bucket, I noticed that the linked bucket under ```/s3v``` still retains the owner field.

Here are the commands I used, executed via Docker Compose under ```ozone/hadoop-ozone/dist/target/ozone-2.1.0-SNAPSHOT/compose/ozone```:
```sh
ozone sh volume create /test

ozone sh volume create /s3v

ozone sh bucket create /test/source

ozone sh bucket link /test/source /s3v/link

ozone sh bucket delete /test/source

# owner still exists
ozone sh bucket info /s3v/link
aws s3api get-bucket-acl --bucket link --endpoint-url http://s3g:9878
```

Is this the expected behavior?

ivandika3 replied:
Thanks for testing it.

Yes, seems `RpcClient#createBucket` will populate the owner (and other with the current user (`UserGroupInformation#getShortUserName`) if the owner is not specified in the `BucketArgs` (e.g. when creating a link bucket or when bucket creation does not specify owner name).

This might cause some inconsistencies between the link bucket and the source bucket info (e.g. different owner, different bucket layout). Currently most / if not all the operations on link bucket should be sent to the source bucket since both are referring to the same bucket. So technically a link bucket should only contain enough info to point to the source bucket.

I think for now we can simply follow the current behavior that dangling bucket can have an owner. In the future, we might need to modify `getBucketInfo` when handling link bucket.

hevinhsu replied:
Thanks for the explanation! I’ll test it based on the current behavior.

peterxcli commented:
Sure.

hevinhsu commented:
Hi @ivandika3 @peterxcli ,

I've made some changes, including:

* Moved static methods  in```BucketOwnerCondition```  to ```S3Owner```
* Added null/empty checks for owner fields for bucketOwner verification
* Added integration tests for LinkBucket, including:
  * Normal link bucket scenario
  * Dangling bucket scenario (in the current version, the owner field is not null)
* Removed unit tests for bucket owner condition, as these scenarios are now covered by the integration tests (let me know if you prefer to keep the unit tests, and I can add them back)

Would you mind taking a look?
Thanks

ivandika3 commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:
```diff
@@ -889,6 +898,9 @@ public Response completeMultipartUpload(@PathParam("bucket") String bucket,
 
     OmMultipartUploadCompleteInfo omMultipartUploadCompleteInfo;
     try {
+      OzoneBucket ozoneBucket = volume.getBucket(bucket);
+      S3Owner.verifyBucketOwnerCondition(headers, bucket, ozoneBucket.getOwner());
+
```
Nit: The following `getClientProtocol().completeMultipartUpload` can be changed to use `ozoneBucket#completeMultipartUpload` instead.

This is just a preference and can be ignored. Using `getClientProtocol()` avoids the OM getBucket call, but since it's called anyway, we can use `OzoneBucket`.

ivandika3 commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:
```diff
@@ -1236,6 +1250,10 @@ private CopyObjectResponse copyObject(OzoneVolume volume,
     String sourceKey = result.getRight();
     DigestInputStream sourceDigestInputStream = null;
+
+    String sourceBucketOwner = volume.getBucket(sourceBucket).getOwner();
+    String destBucketOwner = volume.getBucket(destBucket).getOwner();
+    S3Owner.verifyBucketOwnerConditionOnCopyOperation(headers, destBucket, sourceBucketOwner, destBucketOwner);
```
Since the `destBucket` owner is already checked in the outer scope (`ObjectEndpoint#put`), there is no need to call `getBucket` again. We can simply check the source bucket.

ivandika3 reviewed (commented):
Thanks @hevinhsu for the continued effort on this patch and the extensive tests. Overall looks good. I left some minor comments (mostly some nits), after they are addressed, I think this patch is good to go.

The main follow up we need to do is reduce the additional `getBucket` OM call introduced in this patch. In addition to the comments, we can think about a way to only call `getBucket` if the `x-amz-expected-bucket-owner` or `x-amz-source-expected-bucket-owner` is specified so that normal requests without expected bucket owner won't see increased latency. This would make the code slightly more complicated, but would save us some RTT.

ivandika3 commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/S3Owner.java:
```diff
@@ -80,4 +85,57 @@ public String toString() {
+   * @param destOwner     destination bucket owner
+   * @throws OS3Exception if the expected source or destination bucket owner does not match
+   */
+  public static void verifyBucketOwnerConditionOnCopyOperation(HttpHeaders headers, String bucketName,
+                                                               String sourceOwner, String destOwner)
```
Since `copyObject` can be done across bucket, we need to specify both the `destBucketName` and `sourceBucketName` and use that in the `S3ErrorTable.newError(S3ErrorTable.BUCKET_OWNER_MISMATCH`)

ivandika3 reviewed (commented):
@hevinhsu Thanks for the update. One more comment I missed.

hevinhsu commented:
@ivandika3 Thanks again for the great suggestions — most of them have now been addressed!

I initially chose to use a static method for bucket owner validation because many APIs already require an OzoneBucket, and I felt this approach would minimize the amount of changes needed in the current codebase.

That said, I really appreciate your point about reducing the getBucket call to improve RTT. I'll give it some thought and explore how we might optimize this further.

As for filing a follow-up issue, I'm not quite sure how to scope it yet to ensure the changes stay manageable. If you have s

ivandika3 reviewed (approved):
Thanks for the improvements and updates so far. LGTM +1.

> As for filing a follow-up issue, I'm not quite sure how to scope it yet to ensure the changes stay manageable.

We can simply raise a ticket to reduce the `getBucket` info first. We can split them into multiple tasks to make them more manageable if needed.

hevinhsu commented:
Thanks for the review!

I've created the follow-up Jira issue here: [HDDS-13270](https://issues.apache.org/jira/browse/HDDS-13270).
Feel free to update it directly or let me know if there’s anything that needs to be adjusted or added.

peterxcli commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/S3Owner.java:
```diff
@@ -80,4 +85,59 @@ public String toString() {
+   * @param bucketOwner   bucket owner
+   * @throws OS3Exception if the expected bucket owner does not match
+   */
+  public static void verifyBucketOwnerCondition(HttpHeaders headers, String bucketName, String bucketOwner)
+      throws OS3Exception {
+    if (headers == null || bucketOwner == null) {
+      return;
+    }
+
+    final String expectedBucketOwner = headers.getHeaderString(S3Consts.EXPECTED_BUCKET_OWNER_HEADER);
+
+    if (StringUtils.isEmpty(expectedBucketOwner)) {
+      return;
+    }
+    if (expectedBucketOwner.equals(bucketOwner)) {
+      return;
+    }
+    throw S3ErrorTable.newError(S3ErrorTable.BUCKET_OWNER_MISMATCH, bucketName);
+  }
```
nit:

```suggestion
  public static void verifyBucketOwnerCondition(HttpHeaders headers, String bucketName, String headerKey, String actualOwner)
    throws OS3Exception {
	  if (headers == null || actualOwner == null) {
	    return;
	  }
	  final String expectedOwner = headers.getHeaderString(headerKey);
	  if (StringUtils.isEmpty(expectedOwner)) {
	    return;
	  }
	  if (!expectedOwner.equals(actualOwner)) {
	    throw S3ErrorTable.newError(S3ErrorTable.BUCKET_OWNER_MISMATCH, bucketName);
	  }
  }
  
  public static void verifyBucketOwnerCondition(HttpHeaders headers, String bucketName, String actualOwner) throws OS3Exception {
    verifyBucketOwnerCondition(headers, bucketName, S3Consts.EXPECTED_BUCKET_OWNER_HEADER, actualOwner);
}

```

peterxcli commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/S3Owner.java:
```diff
@@ -80,4 +85,59 @@ public String toString() {
+   * @param destOwner        destination bucket owner
+   * @throws OS3Exception if the expected source or destination bucket owner does not match
+   */
+  public static void verifyBucketOwnerConditionOnCopyOperation(HttpHeaders headers, String sourceBucketName,
+                                                               String sourceOwner, String destBucketName,
+                                                               String destOwner)
+      throws OS3Exception {
+    if (headers == null) {
+      return;
+    }
+
+    final String expectedSourceOwner = headers.getHeaderString(S3Consts.EXPECTED_SOURCE_BUCKET_OWNER_HEADER);
+    final String expectedDestOwner = headers.getHeaderString(S3Consts.EXPECTED_BUCKET_OWNER_HEADER);
+
+    if (expectedSourceOwner != null && sourceOwner != null && !sourceOwner.equals(expectedSourceOwner)) {
+      throw S3ErrorTable.newError(S3ErrorTable.BUCKET_OWNER_MISMATCH, sourceBucketName);
+    }
+
+    if (expectedDestOwner != null && destOwner != null && !destOwner.equals(expectedDestOwner)) {
+      throw S3ErrorTable.newError(S3ErrorTable.BUCKET_OWNER_MISMATCH, destBucketName);
+    }
+  }
```
nit:
```suggestion
  public static void verifyBucketOwnerConditionOnCopyOperation(HttpHeaders headers, String sourceBucketName, String sourceOwner, String destBucketName, String destOwner) throws OS3Exception {
    if (headers == null) {
      return;
    }
	verifyBucketOwnerCondition(headers, sourceBucketName, S3Consts.EXPECTED_SOURCE_BUCKET_OWNER_HEADER, sourceOwner);
	verifyBucketOwnerCondition(headers, destBucketName, S3Consts.EXPECTED_BUCKET_OWNER_HEADER, destOwner);
}
```

peterxcli reviewed (approved):
Thanks @hevinhsu updating the patch. LGTM! 
I have two minor suggestion about simplifying the logic in S3Owner, but I think they can be addressed in follow-up, too.

peterxcli commented:
Merged. Thanks @hevinhsu and @ivandika3 for the continuous effort!

hevinhsu commented:
Thanks @ivandika3 @peterxcli for your suggestions and review.
